micrometer-core/src/main/java/io/micrometer/core/instrument/push/PushMeterRegistry.java

```diff
@@ -97,9 +103,8 @@ public void start(ThreadFactory threadFactory) {
                     + TimeUtils.format(config.step()));
 
             scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(threadFactory);
-            // time publication to happen just after StepValue finishes the step
             long stepMillis = config.step().toMillis();
-            long initialDelayMillis = stepMillis - (clock.wallTime() % stepMillis) + 1;
+            long initialDelayMillis = calculateInitialDelay();
```

Comment:
This in a way enforces registries to complete the publish() in 20% of the step time. And with the issues referenced in https://github.com/micrometer-metrics/micrometer/issues/3717, this could be a serious issue for users, unless there is an option to preserve the existing behavior.

If the step registries can publish within 20% of the step time, then I believe that alone would ensure that we are reading the data for right step.

Reply:
If you aren't publishing within 20% of the step time, that's probably a separate issue to look into. I would think there must be something wrong with the registry implementation, high cardinality, not enough CPU/memory, or not enough resources for your metrics backend. Prior to Micrometer 1.7, we published at any arbitrary time in the step, and I don't think anyone reported publishing taking so long. If publishing takes a long time, I don't think the answer is to publish at the beginning of the step to give the most time possible. That just exacerbates the issue because it will slow down shared resources (network, metrics backend) when all of your instances try to use them at the same time.

> If the step registries can publish within 20% of the step time, then I believe that alone would ensure that we are reading the data for right step.

Not if we're scheduling at the end of the step. That's the whole point. Registries still need to publish in a reasonable amount of time, and we need to look into cases where that isn't happening.

Reply:
> Not if we're scheduling at the end of the step.

I agree with this. But, what I meant is if 20% is reasonable enough time to publish we can do the publishing anytime. But, yeah that will bring the issue of Step's not reporting data for the particular step. Fair enough.

But, I would probably have some config to have the old behaviour for at least next 1 or 2 versions. I will leave you guys to decide on that.

Reply:
We try very hard to not break compatibility, which means it is very hard for us to remove something once exposed as public API. This is one of the reasons I try very hard to not make things configurable that in theory we believe users should not need to adjust. I would like to get more feedback on this with RC1 and we can consider making it configurable in GA or the next release if there's a need. Let's look into it more.